File: src/main/python/org/o3project/odenos/remoteobject/manager/component/component_manager.py
# -*- coding:utf-8 -*-

# Copyright 2015 NEC Corporation.                                          #
#                                                                          #
# Licensed under the Apache License, Version 2.0 (the "License");          #
# you may not use this file except in compliance with the License.         #
# You may obtain a copy of the License at                                  #
#                                                                          #
#   http://www.apache.org/licenses/LICENSE-2.0                             #
#                                                                          #
# Unless required by applicable law or agreed to in writing, software      #
# distributed under the License is distributed on an "AS IS" BASIS,        #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. #
# See the License for the specific language governing permissions and      #
# limitations under the License.                                           #

import logging
import copy
from urlparse import urlparse

from org.o3project.odenos.remoteobject.remote_object import RemoteObject
from org.o3project.odenos.remoteobject.object_property import ObjectProperty
from org.o3project.odenos.remoteobject.message.request import Request
from org.o3project.odenos.remoteobject.message.response import Response
from org.o3project.odenos.remoteobject.manager.component.event.component_changed import(
    ComponentChanged
)
from org.o3project.odenos.remoteobject.manager.component.component_type\
    import ComponentType
from org.o3project.odenos.remoteobject.manager.system.event.component_manager_changed\
    import ComponentManagerChanged
from org.o3project.odenos.core.util.request_parser import RequestParser


class ComponentManager(RemoteObject):

    # FIXME
    DESCRIPTION = "ComponentManager for python"

    def __init__(self, object_id, dispatcher):
        RemoteObject.__init__(self, object_id, dispatcher)
        self.component_classes = {}
        self.components = {}
        self.dispatcher = dispatcher
        self.__parser = RequestParser()
        self.__add_rules()
        self._object_property.set_state(ObjectProperty.State.RUNNING)

    def register_to_system_manager(self):
        logging.debug("object_property of ComponentManager %s is %s",
                      self.object_id,
                      self._object_property.packed_object())

        self.__register_component_managers()
        path = "component_managers/%s" % self.object_id
        resp = self._request(self.dispatcher.system_manager_id,
                             Request.Method.PUT,
                             path,
                             self._object_property)
        if resp.is_error(Request.Method.PUT):
            logging.error("Failed registration to SystemManager.")
            self._object_property.set_state(ObjectProperty.State.ERROR)
            return

        self.__register_event_manager()
        self.__subscribe_event()
        logging.info("Complete registration to SystemManager.")

    def __register_component_managers(self):
        resp = self._request(self.dispatcher.system_manager_id,
                             Request.Method.GET,
                             "component_managers",
                             None)
        if resp.is_error(Request.Method.GET):
            logging.error("Failed get component_managers from SystemManager.")
            self._object_property.set_state(ObjectProperty.State.ERROR)
            return

        for component_manager in resp.body:
            self.__register_other_component_manager(component_manager)

    def __register_other_component_manager(self, component_manager):
        object_id = component_manager[ObjectProperty.OBJECT_ID]
        if object_id == self.object_id:
            return

        logging.info("Register Other Component Manager ID:%s", object_id)
        self.dispatcher.add_remote_client(object_id)

    def __unregister_component_manager(self, object_id):
        self.dispatcher.remove_remote_client(object_id)

    def __register_event_manager(self):
        resp = self._request(self.dispatcher.system_manager_id,
                             Request.Method.GET,
                             "objects/%s" % self.dispatcher.event_manager_id,
                             None)
        if resp.is_error(Request.Method.GET):
            self._object_property.set_state(ObjectProperty.State.ERROR)
            return

        self.dispatcher.add_remote_client(self.dispatcher.event_manager_id)

    def __subscribe_event(self):
        self._event_subscription.add_filter(
            self.dispatcher.system_manager_id,
            ComponentManagerChanged.TYPE)
        self._apply_event_subscription()

    def register_component_type(self, component):
        component_name = component.__name__
        logging.info("Register Component Type:%s", component_name)

        self.component_classes[component_name] = component
        component_types = \
            self._object_property.get_property(ObjectProperty.COMPONENT_TYPES)
        if component_types:
            component_types += (",%s" % component_name)
        else:
            component_types = "%s" % component_name
        self._object_property.set_property(ObjectProperty.COMPONENT_TYPES,
                                           component_types)

    def __add_rules(self):
        rules = []
        rules.append({RequestParser.PATTERN: r"^component_types/?$",
                      RequestParser.METHOD: Request.Method.GET,
                      RequestParser.FUNC: self._do_get_component_types,
                      RequestParser.PARAMS: 0})
        rules.append({RequestParser.PATTERN: r"^components/?$",
                      RequestParser.METHOD: Request.Method.GET,
                      RequestParser.FUNC: self._do_get_components,
                      RequestParser.PARAMS: 0})
        rules.append({RequestParser.PATTERN: r"^components/"
                      + "([a-zA-Z0-9_-]+)/?$",
                      RequestParser.METHOD: Request.Method.PUT,
                      RequestParser.FUNC: self._do_put_component,
                      RequestParser.PARAMS: 2})
        rules.append({RequestParser.PATTERN: r"^components/"
                      + "([a-zA-Z0-9_-]+)/?$",
                      RequestParser.METHOD: Request.Method.GET,
                      RequestParser.FUNC: self._do_get_component,
                      RequestParser.PARAMS: 1})
        rules.append({RequestParser.PATTERN: r"^components/"
                      + "([a-zA-Z0-9_-]+)/?$",
                      RequestParser.METHOD: Request.Method.DELETE,
                      RequestParser.FUNC: self._do_delete_component,
                      RequestParser.PARAMS: 1})
        self.__parser.add_rule(rules)

    def _on_request(self, request):
        return self.__parser.action(request)

    def _do_get_component(self, object_id):
        if object_id in self.components:
            return Response(Response.StatusCode.OK,
                            self.components[object_id].object_property)
        return Response(Response.StatusCode.NOT_FOUND, None)

    def _do_put_component(self, obj_prop, object_id):
        component_type = obj_prop[ObjectProperty.OBJECT_TYPE]

        if component_type not in self.component_classes:
            return Response(Response.StatusCode.BAD_REQUEST, None)
        elif object_id in self.components:
            return Response(Response.StatusCode.CONFLICT, None)

        component_class = self.component_classes[component_type]
        self.components[object_id] = component_class(object_id,
                                                     self.dispatcher)
        if self.components[object_id].on_initialize(obj_prop):
            self.components[object_id].\
                _object_property.set_state(ObjectProperty.State.RUNNING)
        else:
            self.components[object_id].\
                _object_property.set_state(ObjectProperty.State.ERROR)
        curr = self.components[object_id].object_property.packed_object()
        self._do_component_changed(ComponentChanged.Action.ADD,
                                   None,
                                   curr)

        logging.info("Created Component Type:%s ID:%s",
                     component_type, object_id)
        return Response(Response.StatusCode.CREATED,
                        self.components[object_id].object_property)

    def _do_delete_component(self, object_id):
        if object_id in self.components:
            component = self.components[object_id]
            prev = copy.deepcopy(component._object_property).packed_object()
            component.on_finalize()
            del self.components[object_id]
            self._do_component_changed(ComponentChanged.Action.DELETE,
                                       prev,
                                       None)
            logging.info("Deleted Component ID:%s", object_id)

        return Response(Response.StatusCode.OK, None)

    def _do_get_components(self):
        body = {}
        for object_id in self.components:
            body[object_id] = \
                self.components[object_id].object_property.packed_object()
        return Response(Response.StatusCode.OK, body)

    def _do_get_component_types(self):
        comp_types = {}
        try:
            for type_name, clazz in self.component_classes.items():
                comp_id = "%s_%s" % (self.object_id, type_name)
                component = clazz(comp_id, None)
                obj_prop = component.object_property

                type = obj_prop.get_property(ObjectProperty.OBJECT_TYPE)
                super_type = obj_prop.get_property(ObjectProperty.OBJECT_SUPER_TYPE)

                connection_types = {}
                connection_types_str = obj_prop.get_property(
                    ObjectProperty.CONNECTION_TYPES)
                conn_type_list = connection_types_str.split(",")
                for type_elem in conn_type_list:
                    type_elem_list = type_elem.split(":")
                    if len(type_elem_list) == 2:
                        connection_types[type_elem_list[0]] = type_elem_list[1]

                description = obj_prop.get_property(ObjectProperty.DESCRIPTION)

                target = ComponentType(type, super_type,
                                       connection_types, description)
                comp_types[type_name] = target.packed_object()

        except Exception, e:
            return Response(Response.StatusCode.INTERNAL_SERVER_ERROR,
                            str(e))

        return Response(Response.StatusCode.OK, comp_types)

    def _do_component_changed(self, action, prev, curr):
        body = {ComponentChanged.ACTION: action,
                ComponentChanged.PREV: prev,
                ComponentChanged.CURR: curr}
        self._publish_event_async(ComponentChanged.TYPE, body)

    def _do_event_componentmanagerchanged(self, event):
        msg = None
        try:
            msg = ComponentManagerChanged.create_from_packed(event.body)
        except KeyError, e:
            logging.error("Receive Invalid ComponentManagerChanged Message"
                          + "KeyError: " + str(e))
            return

        if msg.action == ComponentManagerChanged.Action.ADD:
            self.__register_other_component_manager(msg.curr)
        elif msg.action == ComponentManagerChanged.Action.DELETE:
            self.__unregister_component_manager(
                msg.prev[ObjectProperty.OBJECT_ID])

File: src/main/java/org/o3project/odenos/remoteobject/manager/component/ComponentManager.java
/*
 * Copyright 2015 NEC Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.o3project.odenos.remoteobject.manager.component;

import org.o3project.odenos.core.component.Component;
import org.o3project.odenos.core.component.SystemManagerInterface;
import org.o3project.odenos.core.manager.system.event.ComponentManagerChanged;
import org.o3project.odenos.remoteobject.ObjectProperty;
import org.o3project.odenos.remoteobject.RemoteObject;
import org.o3project.odenos.remoteobject.RequestParser;
import org.o3project.odenos.remoteobject.manager.ObjectPropertyList;
import org.o3project.odenos.remoteobject.manager.component.event.ComponentChanged;
import org.o3project.odenos.remoteobject.message.Event;
import org.o3project.odenos.remoteobject.message.MessageBodyUnpacker.ParseBodyException;
import org.o3project.odenos.remoteobject.message.Request;
import org.o3project.odenos.remoteobject.message.Response;
import org.o3project.odenos.remoteobject.messagingclient.MessageDispatcher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.lang.reflect.Constructor;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

/**
 * ComponentManager class manages components life-cycle.
 *
 */
public class ComponentManager extends RemoteObject {

  private static final Logger log = LoggerFactory.getLogger(ComponentManager.class);

  protected SystemManagerInterface sysMngIf;
  protected Map<String, Class<? extends Component>> componentClasses;
  protected Map<String, Component> components;

  private final RequestParser<IActionCallback> parser;
  private static final String COMP_ID = "comp_id";

  /**
   * Constructor.
   * @param objectId object ID.
   * @param baseUri base URI.
   * @param dispatcher Message Dispatcher object.
   * @deprecated {@link #ComponentManager(String, MessageDispatcher)}
   */
  @Deprecated
  public ComponentManager(String objectId, String baseUri,
      MessageDispatcher dispatcher) {
    this(objectId, dispatcher);
  }

  /**
   * Constructor.
   * @param objectId object ID.
   * @param dispatcher Message Dispatcher object.
   */
  public ComponentManager(String objectId, MessageDispatcher dispatcher) {
    super(objectId, dispatcher);
    componentClasses = new HashMap<String, Class<? extends Component>>();
    components = new HashMap<String, Component>();

    this.getProperty().setObjectState(ObjectProperty.State.RUNNING);
    this.sysMngIf = new SystemManagerInterface(dispatcher, objectId);

    parser = this.createParser();
  }

  /**
   * Register to system manager.
   * @throws Exception if an error occurs.
   */
  public void registerToSystemManager() throws Exception {
    Response rsp = this.sysMngIf.putComponentMng(objectProperty);
    if (!rsp.statusCode.equals(Response.OK)) {
      throw new Exception("failed 'PUT component_managers'");
    }

    this.registerEventManager();
    this.subscribeEvents();
    this.registerComponentManagers();
    return;
  }

  private void registerEventManager() throws Exception {
    String eventManagerId = this.getEventManagerId();

    ObjectProperty eventMngObj = this.sysMngIf.getObject(eventManagerId);
    if (eventMngObj == null) {
      log.error("Internal Error to Get objects/" + eventManagerId);
      throw new Exception();
    }

    try {
      this.addRemoteObject(eventManagerId);
    } catch (IOException e) {
      log.error("Failure to addRemoteObject");
      throw new Exception();
    }
    return;
  }

  private void subscribeEvents() throws Exception {
    this.eventSubscription.addFilter(
        this.getSystemManagerId(),
        ComponentManagerChanged.TYPE);
    Response rsp = this.applyEventSubscription();
    if (!rsp.statusCode.equals(Response.OK)) {
      log.error("Can't subscribe the ComponentManagerChanged.");
      throw new Exception();
    }
  }

  private void registerComponentManagers() throws Exception {
    ObjectPropertyList componentManagers = this.sysMngIf.getComponentMngs();
    if (componentManagers == null) {
      log.error("Internal Error to Get component_managers.");
      throw new Exception();
    }
    for (ObjectProperty componentManager : componentManagers) {
      this.registerOtherComponentManager(componentManager);
    }
    return;
  }

  private void registerOtherComponentManager(ObjectProperty componentManager)
      throws Exception {
    if (this.getObjectId().equals(componentManager.getObjectId())) {
      return;
    }
    try {
      this.addRemoteObject(componentManager.getObjectId());
    } catch (IOException e) {
      log.error("Failure to addRemoteObject");
      throw new Exception();
    }
    return;
  }

  private void unregisterComponentManager(String objectId) throws Exception {
    try {
      this.removeRemoteObject(objectId);
    } catch (IOException e) {
      log.error("Failure to unregisterComponentManager");
      throw new Exception();
    }
    return;
  }

  /**
   * Register a type of component.
   * @param component class of component.
   */
  public void registerComponentType(Class<? extends Component> component) {
    String componentType = component.getSimpleName();

    if (!componentClasses.containsKey(componentType)) {
      componentClasses.put(componentType, component);
      String componentTypes = this.objectProperty.getProperty(
          ObjectProperty.PropertyNames.COMPONENT_TYPES);
      if (componentTypes == null) {
        componentTypes = componentType;
      } else {
        componentTypes += "," + componentType;
      }
      this.objectProperty.setProperty(
          ObjectProperty.PropertyNames.COMPONENT_TYPES,
          componentTypes);
    }
  }

  @Override
  protected Response onRequest(Request request) {
    log.debug("onRequest: " + request.method + ", " + request.path);

    RequestParser<IActionCallback>.ParsedRequest parsed = parser
        .parse(request);
    Response response = null;

    IActionCallback callback = parsed.getResult();
    if (callback == null) {
      return new Response(Response.BAD_REQUEST, null);
    }
    try {
      response = callback.process(parsed);
    } catch (Exception e) {
      log.error("Exception Request: " + request.method + ", "
          + request.path);
      response = new Response(Response.BAD_REQUEST, null);
    }
    if (response == null) {
      response = new Response(Response.BAD_REQUEST, null);
    }
    return response;
  }

  private RequestParser<IActionCallback> createParser() {
    return new RequestParser<IActionCallback>() {
      {
        addRule(Request.Method.GET,
            "component_types",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed) {
                return getComponentTypes();
              }
            });
        addRule(Request.Method.GET,
            "components",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed) {
                return getComponents();
              }
            });
        addRule(Request.Method.GET,
            "components/<comp_id>",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed) {
                return getComponentId(parsed.getParam(COMP_ID));
              }
            });
        addRule(Request.Method.PUT,
            "components/<comp_id>",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed)
                  throws ParseBodyException {
                return putComponentId(
                    parsed.getParam(COMP_ID),
                    parsed.getRequest().getBody(ObjectProperty.class));
              }
            });
        addRule(Request.Method.DELETE,
            "components/<comp_id>",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed) {
                return deleteComponentId(parsed
                    .getParam(COMP_ID));
              }
            });
      }
    };
  }

  protected Response getComponentTypes() {
    Map<String, ComponentType> compTypes = new HashMap<>();

    for (String cmType : componentClasses.keySet()) {
      Component component = null;
      String objectId = String.format("%s_%s", this.getObjectId(), cmType);
      try {
        Class<? extends Component> componentClass = componentClasses.get(cmType);
        Constructor<? extends Component> ct = null;
        try {
          ct = componentClass.getConstructor(String.class,
              MessageDispatcher.class);
          component = ct.newInstance(objectId, null);
        } catch (NoSuchMethodException e) {
          ct = componentClass.getConstructor(String.class,
              String.class, MessageDispatcher.class);
          component = ct.newInstance(objectId,
              objectProperty.getBaseUri(), null);
        }
        ObjectProperty objProp = component.getProperty();
        String type = objProp.getProperty(
            ObjectProperty.PropertyNames.OBJECT_TYPE);
        String superType = objProp.getProperty(
            ObjectProperty.PropertyNames.OBJECT_SUPER_TYPE);

        Map<String, String> connectionTypes = new HashMap<>();
        String connectionTypesStr = objProp.getProperty(
            ObjectProperty.PropertyNames.CONNECTION_TYPES);
        String[] connStrList = connectionTypesStr.split(",");
        for (String connTypeElem : connStrList) {
          String[] connTypeElemList = connTypeElem.split(":");
          if (connTypeElemList.length == 2) {
            connectionTypes.put(connTypeElemList[0], connTypeElemList[1]);
          }
        }

        String description = objProp.getProperty(
            ObjectProperty.PropertyNames.DESCRIPTION);
        compTypes.put(type,
            new ComponentType(type, superType, connectionTypes, description));
      } catch (Exception e) {
        return new Response(Response.INTERNAL_SERVER_ERROR, e.getMessage());
      }
    }
    return new Response(Response.OK, compTypes);
  }

  protected Response getComponents() {
    Map<String, ObjectProperty> componentsObj = new HashMap<String, ObjectProperty>();
    for (Entry<String, Component> e : components.entrySet()) {
      componentsObj.put(e.getKey(), e.getValue().getProperty());
    }
    return new Response(Response.OK, componentsObj);
  }

  protected Response getComponentId(String objectId) {
    Component component = components.get(objectId);
    if (component == null) {
      return new Response(Response.NOT_FOUND, null);
    }
    return new Response(Response.OK, component.getProperty());
  }

  protected Response putComponentId(
      String objectId, ObjectProperty prop) {
    if (prop == null) {
      return new Response(Response.BAD_REQUEST,
          "Bad format: ObjectProperty is expected");
    }

    if (!componentClasses.containsKey(prop.getObjectType())) {
      return new Response(Response.BAD_REQUEST, "Error unknown type ");
    }

    if (components.containsKey(objectId)) {
      return new Response(Response.CONFLICT,
          "Component is already created");
    }
    prop.setProperty(
        ObjectProperty.PropertyNames.OBJECT_ID, objectId);

    Component component = null;
    try {
      Class<? extends Component> componentClass = componentClasses
          .get(prop.getObjectType());
      Constructor<? extends Component> ct = null;
      try {
        ct = componentClass.getConstructor(String.class,
            MessageDispatcher.class);
        component = ct.newInstance(objectId, messageDispatcher);
      } catch (NoSuchMethodException e) {
        ct = componentClass.getConstructor(String.class,
            String.class, MessageDispatcher.class);
        component = ct.newInstance(objectId,
            objectProperty.getBaseUri(), messageDispatcher);
      }
      components.put(objectId, component);
      if (component.onInitialize(prop)) {
        component.setState(ObjectProperty.State.RUNNING);
      } else {
        component.setState(ObjectProperty.State.ERROR);
      }
    } catch (Exception e) {
      return new Response(Response.INTERNAL_SERVER_ERROR, e.getMessage());
    }

    componentChanged(ComponentChanged.Action.add.name(), null, prop);

    return new Response(Response.CREATED, component.getProperty());
  }

  protected Response deleteComponentId(String objectId) {
    if (components.containsKey(objectId)) {
      Component component = components.get(objectId);
      ObjectProperty prev = (ObjectProperty) component
          .getProperty().clone();
      component.onFinalize();
      components.remove(objectId);

      componentChanged(ComponentChanged.Action.delete.name(), prev, null);
    }
    return new Response(Response.OK, null);
  }

  protected final void componentChanged(final String action,
      final ObjectProperty prev,
      final ObjectProperty curr) {
    ComponentChanged msg = new ComponentChanged(action, prev, curr);
    try {
      publishEvent(ComponentChanged.TYPE, msg);
    } catch (Exception e) {
      log.error("Failed to ComponentChanged");
    }
  }

  @Override
  protected void onEvent(Event event) {
    if (event.eventType.equals(
        ComponentManagerChanged.TYPE)) {
      try {
        ComponentManagerChanged prop = event
            .getBody(ComponentManagerChanged.class);
        if (prop.action().equals(
            ComponentManagerChanged.Action.add.name())) {
          this.registerOtherComponentManager(prop.curr());
        } else if (prop.action().equals(
            ComponentManagerChanged.Action.delete.name())) {
          this.unregisterComponentManager(prop.prev().getObjectId());
        }
      } catch (Exception e) {
        log.error("Can't register ComponentManager.", e);
      }
    }
  }
}
